// @aws-sdk/client-location/dist-cjs/Location.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Location = void 0;
const smithy_client_1 = require("@smithy/smithy-client");
const AssociateTrackerConsumerCommand_1 = require("./commands/AssociateTrackerConsumerCommand");
const BatchDeleteDevicePositionHistoryCommand_1 = require("./commands/BatchDeleteDevicePositionHistoryCommand");
const BatchDeleteGeofenceCommand_1 = require("./commands/BatchDeleteGeofenceCommand");
const BatchEvaluateGeofencesCommand_1 = require("./commands/BatchEvaluateGeofencesCommand");
const BatchGetDevicePositionCommand_1 = require("./commands/BatchGetDevicePositionCommand");
const BatchPutGeofenceCommand_1 = require("./commands/BatchPutGeofenceCommand");
const BatchUpdateDevicePositionCommand_1 = require("./commands/BatchUpdateDevicePositionCommand");
const CalculateRouteCommand_1 = require("./commands/CalculateRouteCommand");
const CalculateRouteMatrixCommand_1 = require("./commands/CalculateRouteMatrixCommand");
const CreateGeofenceCollectionCommand_1 = require("./commands/CreateGeofenceCollectionCommand");
const CreateKeyCommand_1 = require("./commands/CreateKeyCommand");
const CreateMapCommand_1 = require("./commands/CreateMapCommand");
const CreatePlaceIndexCommand_1 = require("./commands/CreatePlaceIndexCommand");
const CreateRouteCalculatorCommand_1 = require("./commands/CreateRouteCalculatorCommand");
const CreateTrackerCommand_1 = require("./commands/CreateTrackerCommand");
const DeleteGeofenceCollectionCommand_1 = require("./commands/DeleteGeofenceCollectionCommand");
const DeleteKeyCommand_1 = require("./commands/DeleteKeyCommand");
const DeleteMapCommand_1 = require("./commands/DeleteMapCommand");
const DeletePlaceIndexCommand_1 = require("./commands/DeletePlaceIndexCommand");
const DeleteRouteCalculatorCommand_1 = require("./commands/DeleteRouteCalculatorCommand");
const DeleteTrackerCommand_1 = require("./commands/DeleteTrackerCommand");
const DescribeGeofenceCollectionCommand_1 = require("./commands/DescribeGeofenceCollectionCommand");
const DescribeKeyCommand_1 = require("./commands/DescribeKeyCommand");
const DescribeMapCommand_1 = require("./commands/DescribeMapCommand");
const DescribePlaceIndexCommand_1 = require("./commands/DescribePlaceIndexCommand");
const DescribeRouteCalculatorCommand_1 = require("./commands/DescribeRouteCalculatorCommand");
const DescribeTrackerCommand_1 = require("./commands/DescribeTrackerCommand");
const DisassociateTrackerConsumerCommand_1 = require("./commands/DisassociateTrackerConsumerCommand");
const GetDevicePositionCommand_1 = require("./commands/GetDevicePositionCommand");
const GetDevicePositionHistoryCommand_1 = require("./commands/GetDevicePositionHistoryCommand");
const GetGeofenceCommand_1 = require("./commands/GetGeofenceCommand");
const GetMapGlyphsCommand_1 = require("./commands/GetMapGlyphsCommand");
const GetMapSpritesCommand_1 = require("./commands/GetMapSpritesCommand");
const GetMapStyleDescriptorCommand_1 = require("./commands/GetMapStyleDescriptorCommand");
const GetMapTileCommand_1 = require("./commands/GetMapTileCommand");
const GetPlaceCommand_1 = require("./commands/GetPlaceCommand");
const ListDevicePositionsCommand_1 = require("./commands/ListDevicePositionsCommand");
const ListGeofenceCollectionsCommand_1 = require("./commands/ListGeofenceCollectionsCommand");
const ListGeofencesCommand_1 = require("./commands/ListGeofencesCommand");
const ListKeysCommand_1 = require("./commands/ListKeysCommand");
const ListMapsCommand_1 = require("./commands/ListMapsCommand");
const ListPlaceIndexesCommand_1 = require("./commands/ListPlaceIndexesCommand");
const ListRouteCalculatorsCommand_1 = require("./commands/ListRouteCalculatorsCommand");
const ListTagsForResourceCommand_1 = require("./commands/ListTagsForResourceCommand");
const ListTrackerConsumersCommand_1 = require("./commands/ListTrackerConsumersCommand");
const ListTrackersCommand_1 = require("./commands/ListTrackersCommand");
const PutGeofenceCommand_1 = require("./commands/PutGeofenceCommand");
const SearchPlaceIndexForPositionCommand_1 = require("./commands/SearchPlaceIndexForPositionCommand");
const SearchPlaceIndexForSuggestionsCommand_1 = require("./commands/SearchPlaceIndexForSuggestionsCommand");
const SearchPlaceIndexForTextCommand_1 = require("./commands/SearchPlaceIndexForTextCommand");
const TagResourceCommand_1 = require("./commands/TagResourceCommand");
const UntagResourceCommand_1 = require("./commands/UntagResourceCommand");
const UpdateGeofenceCollectionCommand_1 = require("./commands/UpdateGeofenceCollectionCommand");
const UpdateKeyCommand_1 = require("./commands/UpdateKeyCommand");
const UpdateMapCommand_1 = require("./commands/UpdateMapCommand");
const UpdatePlaceIndexCommand_1 = require("./commands/UpdatePlaceIndexCommand");
const UpdateRouteCalculatorCommand_1 = require("./commands/UpdateRouteCalculatorCommand");
const UpdateTrackerCommand_1 = require("./commands/UpdateTrackerCommand");
const LocationClient_1 = require("./LocationClient");
const commands = {
    AssociateTrackerConsumerCommand: AssociateTrackerConsumerCommand_1.AssociateTrackerConsumerCommand,
    BatchDeleteDevicePositionHistoryCommand: BatchDeleteDevicePositionHistoryCommand_1.BatchDeleteDevicePositionHistoryCommand,
    BatchDeleteGeofenceCommand: BatchDeleteGeofenceCommand_1.BatchDeleteGeofenceCommand,
    BatchEvaluateGeofencesCommand: BatchEvaluateGeofencesCommand_1.BatchEvaluateGeofencesCommand,
    BatchGetDevicePositionCommand: BatchGetDevicePositionCommand_1.BatchGetDevicePositionCommand,
    BatchPutGeofenceCommand: BatchPutGeofenceCommand_1.BatchPutGeofenceCommand,
    BatchUpdateDevicePositionCommand: BatchUpdateDevicePositionCommand_1.BatchUpdateDevicePositionCommand,
    CalculateRouteCommand: CalculateRouteCommand_1.CalculateRouteCommand,
    CalculateRouteMatrixCommand: CalculateRouteMatrixCommand_1.CalculateRouteMatrixCommand,
    CreateGeofenceCollectionCommand: CreateGeofenceCollectionCommand_1.CreateGeofenceCollectionCommand,
    CreateKeyCommand: CreateKeyCommand_1.CreateKeyCommand,
    CreateMapCommand: CreateMapCommand_1.CreateMapCommand,
    CreatePlaceIndexCommand: CreatePlaceIndexCommand_1.CreatePlaceIndexCommand,
    CreateRouteCalculatorCommand: CreateRouteCalculatorCommand_1.CreateRouteCalculatorCommand,
    CreateTrackerCommand: CreateTrackerCommand_1.CreateTrackerCommand,
    DeleteGeofenceCollectionCommand: DeleteGeofenceCollectionCommand_1.DeleteGeofenceCollectionCommand,
    DeleteKeyCommand: DeleteKeyCommand_1.DeleteKeyCommand,
    DeleteMapCommand: DeleteMapCommand_1.DeleteMapCommand,
    DeletePlaceIndexCommand: DeletePlaceIndexCommand_1.DeletePlaceIndexCommand,
    DeleteRouteCalculatorCommand: DeleteRouteCalculatorCommand_1.DeleteRouteCalculatorCommand,
    DeleteTrackerCommand: DeleteTrackerCommand_1.DeleteTrackerCommand,
    DescribeGeofenceCollectionCommand: DescribeGeofenceCollectionCommand_1.DescribeGeofenceCollectionCommand,
    DescribeKeyCommand: DescribeKeyCommand_1.DescribeKeyCommand,
    DescribeMapCommand: DescribeMapCommand_1.DescribeMapCommand,
    DescribePlaceIndexCommand: DescribePlaceIndexCommand_1.DescribePlaceIndexCommand,
    DescribeRouteCalculatorCommand: DescribeRouteCalculatorCommand_1.DescribeRouteCalculatorCommand,
    DescribeTrackerCommand: DescribeTrackerCommand_1.DescribeTrackerCommand,
    DisassociateTrackerConsumerCommand: DisassociateTrackerConsumerCommand_1.DisassociateTrackerConsumerCommand,
    GetDevicePositionCommand: GetDevicePositionCommand_1.GetDevicePositionCommand,
    GetDevicePositionHistoryCommand: GetDevicePositionHistoryCommand_1.GetDevicePositionHistoryCommand,
    GetGeofenceCommand: GetGeofenceCommand_1.GetGeofenceCommand,
    GetMapGlyphsCommand: GetMapGlyphsCommand_1.GetMapGlyphsCommand,
    GetMapSpritesCommand: GetMapSpritesCommand_1.GetMapSpritesCommand,
    GetMapStyleDescriptorCommand: GetMapStyleDescriptorCommand_1.GetMapStyleDescriptorCommand,
    GetMapTileCommand: GetMapTileCommand_1.GetMapTileCommand,
    GetPlaceCommand: GetPlaceCommand_1.GetPlaceCommand,
    ListDevicePositionsCommand: ListDevicePositionsCommand_1.ListDevicePositionsCommand,
    ListGeofenceCollectionsCommand: ListGeofenceCollectionsCommand_1.ListGeofenceCollectionsCommand,
    ListGeofencesCommand: ListGeofencesCommand_1.ListGeofencesCommand,
    ListKeysCommand: ListKeysCommand_1.ListKeysCommand,
    ListMapsCommand: ListMapsCommand_1.ListMapsCommand,
    ListPlaceIndexesCommand: ListPlaceIndexesCommand_1.ListPlaceIndexesCommand,
    ListRouteCalculatorsCommand: ListRouteCalculatorsCommand_1.ListRouteCalculatorsCommand,
    ListTagsForResourceCommand: ListTagsForResourceCommand_1.ListTagsForResourceCommand,
    ListTrackerConsumersCommand: ListTrackerConsumersCommand_1.ListTrackerConsumersCommand,
    ListTrackersCommand: ListTrackersCommand_1.ListTrackersCommand,
    PutGeofenceCommand: PutGeofenceCommand_1.PutGeofenceCommand,
    SearchPlaceIndexForPositionCommand: SearchPlaceIndexForPositionCommand_1.SearchPlaceIndexForPositionCommand,
    SearchPlaceIndexForSuggestionsCommand: SearchPlaceIndexForSuggestionsCommand_1.SearchPlaceIndexForSuggestionsCommand,
    SearchPlaceIndexForTextCommand: SearchPlaceIndexForTextCommand_1.SearchPlaceIndexForTextCommand,
    TagResourceCommand: TagResourceCommand_1.TagResourceCommand,
    UntagResourceCommand: UntagResourceCommand_1.UntagResourceCommand,
    UpdateGeofenceCollectionCommand: UpdateGeofenceCollectionCommand_1.UpdateGeofenceCollectionCommand,
    UpdateKeyCommand: UpdateKeyCommand_1.UpdateKeyCommand,
    UpdateMapCommand: UpdateMapCommand_1.UpdateMapCommand,
    UpdatePlaceIndexCommand: UpdatePlaceIndexCommand_1.UpdatePlaceIndexCommand,
    UpdateRouteCalculatorCommand: UpdateRouteCalculatorCommand_1.UpdateRouteCalculatorCommand,
    UpdateTrackerCommand: UpdateTrackerCommand_1.UpdateTrackerCommand,
};
class Location extends LocationClient_1.LocationClient {
}
exports.Location = Location;
(0, smithy_client_1.createAggregatedClient)(commands, Location);
